File: Sources/Scratchpad/App/AppEnvironment.swift
import AppKit
import Observation

/// Constructs every protocol-typed dependency via initializer injection and
/// retains the object graph for the lifetime of the app. This is the single
/// place where concrete implementations are chosen; every other type in the
/// app depends only on protocols.
@MainActor
final class AppEnvironment {
    private let settingsStore: SettingsStore
    private let overlayCoordinator: OverlayCoordinator
    private let hotkeyMonitor: HotkeyMonitoring
    private let statusBarController: StatusBarController

    init() {
        let drawingRepository = FileDrawingRepository()
        let settingsStore = SettingsStore()
        let drawingStoreFactory = DrawingStoreFactory(repository: drawingRepository)

        let overlayCoordinator = OverlayCoordinator(
            overlayState: OverlayState(),
            drawingStoreFactory: drawingStoreFactory,
            renderer: CAShapeLayerStrokeRenderer(),
            animator: NSAnimationContextFadeAnimator(),
            settingsStore: settingsStore,
            screenObserving: NSScreenObserver()
        )

        let launchAtLoginService = SMAppServiceLaunchAtLoginService()
        let preferencesWindowController = PreferencesWindowController(
            settingsStore: settingsStore,
            launchAtLoginService: launchAtLoginService
        )
        let statusBarController = StatusBarController(
            overlayCoordinator: overlayCoordinator,
            launchAtLoginService: launchAtLoginService,
            preferencesWindowController: preferencesWindowController,
            aboutPanelPresenter: AboutPanelPresenter()
        )

        self.settingsStore = settingsStore
        self.overlayCoordinator = overlayCoordinator
        self.hotkeyMonitor = CarbonGlobalHotkeyMonitor()
        self.statusBarController = statusBarController

        registerHotkey()
        observeShortcutChanges()
    }

    private func registerHotkey() {
        try? hotkeyMonitor.startMonitoring(settingsStore.settings.globalShortcut) { [weak self] in
            self?.overlayCoordinator.toggle()
        }
    }

    /// `SettingsStore` is `@Observable`, not Combine-based, so reacting to a
    /// live shortcut change from outside SwiftUI means re-subscribing after
    /// every change -- the standard pattern for observing `@Observable`
    /// state imperatively.
    private func observeShortcutChanges() {
        withObservationTracking {
            _ = settingsStore.settings.globalShortcut
        } onChange: { [weak self] in
            Task { @MainActor in
                self?.registerHotkey()
                self?.observeShortcutChanges()
            }
        }
    }
}

File: Windows/Scratchpad/App/AppEnvironment.cs
using Scratchpad.Animations;
using Scratchpad.Drawing;
using Scratchpad.Hotkeys;
using Scratchpad.Overlay;
using Scratchpad.Persistence;
using Scratchpad.Preferences;
using Scratchpad.Rendering;
using Scratchpad.TrayIcon;
using Scratchpad.Utilities;

namespace Scratchpad.App;

/// Constructs every dependency via constructor injection and retains the
/// object graph for the lifetime of the app. The single place where
/// concrete implementations are chosen; every other type in the app depends
/// only on interfaces. Mirrors the macOS app's `AppEnvironment`.
public sealed class AppEnvironment
{
    private readonly SettingsStore _settingsStore;
    private readonly OverlayCoordinator _overlayCoordinator;
    private readonly IHotkeyMonitoring _hotkeyMonitor;
    private readonly TrayIconController _trayIconController;
    private KeyCombo? _registeredCombo;

    public AppEnvironment()
    {
        var drawingRepository = new FileDrawingRepository();
        var settingsStore = new SettingsStore();
        var drawingStoreFactory = new DrawingStoreFactory(drawingRepository);

        var overlayCoordinator = new OverlayCoordinator(
            new OverlayState(),
            drawingStoreFactory,
            new PathStrokeRenderer(),
            new WpfFadeAnimator(),
            settingsStore,
            new WinFormsScreenObserver());

        var launchAtLoginService = new RegistryLaunchAtLoginService();
        var preferencesWindowController = new PreferencesWindowController(settingsStore, launchAtLoginService);
        var trayIconController = new TrayIconController(
            overlayCoordinator,
            launchAtLoginService,
            preferencesWindowController,
            showAbout: () => new AboutWindow().Show());

        _settingsStore = settingsStore;
        _overlayCoordinator = overlayCoordinator;
        _hotkeyMonitor = new Win32GlobalHotkeyMonitor();
        _trayIconController = trayIconController;

        RegisterHotkey();
        // Reacting to a live shortcut change from Preferences means
        // re-registering the OS-level hotkey after every settings save, not
        // just shortcut changes -- `SettingsStore.Settings` has no
        // per-field change notification. `_registeredCombo` skips the
        // re-registration when the shortcut itself did not actually
        // change, since a WPF `Slider` fires `ValueChanged` continuously
        // while dragging (e.g. adjusting pencil thickness), and tearing
        // down/recreating the Win32 hotkey registration on every one of
        // those events would be wasteful and could momentarily leave the
        // hotkey unregistered.
        _settingsStore.PropertyChanged += (_, _) => RegisterHotkey();
    }

    private void RegisterHotkey()
    {
        var combo = _settingsStore.Settings.GlobalShortcut;
        if (_registeredCombo == combo) return;
        _registeredCombo = combo;
        try
        {
            _hotkeyMonitor.StartMonitoring(combo, () => _overlayCoordinator.Toggle());
        }
        catch (HotkeyRegistrationException)
        {
            // Best-effort: if the combo is already claimed by another app,
            // the user can pick a different one in Preferences. Mirrors the
            // macOS app's `try?` around the same call.
        }
    }
}
